src/rust/engine/process_execution/src/local.rs

```diff
@@ -651,8 +653,16 @@ pub async fn prepare_workdir(
       ..WorkunitMetadata::default()
     },
     |_workunit| async move {
+      // Merge reusable input digests into `input_files`.
+      // TODO: Symlink or bind mount reusable inputs into the input root from a persistent
+      //   cache (which would occur after materializing the input files).
+      let merged_input_digest =
+        merge_reusable_input_digests(&store2, input_files, reusable_input_digests)
+          .map_err(|err| format!("Unable to merge digest: {:?}", err))
+          .boxed()
+          .await?;
```

Comment:
Doing this merging in each Runner is error prone: in particular, this change currently does not include the new digest in the remexec digest for the process/action, which is calculated here: https://github.com/pantsbuild/pants/blob/83404c695fc8c2bd66e63fbc58f7159394a9340e/src/rust/engine/process_execution/src/lib.rs#L574-L596

... and even if all Runners are updated to merge, they will re-merge for every attempt to execute.

----

A less error prone approach might be to give `Process` a new calculated field for "the complete input_digest", which would force constructors of `Process`es to execute the merge themselves. Alternatively, add a high level constructor (or builder) that computed the field. There have been [some steps toward adding the builder pattern to `Process`](https://github.com/pantsbuild/pants/blob/83404c695fc8c2bd66e63fbc58f7159394a9340e/src/rust/engine/process_execution/src/lib.rs#L266-L302), but unfortunately they only focused on field optionality, and so they take `&mut Process` rather than being on a separate builder... sorry about that: lesson learned.

Reply:
To unblock #13787, I'm going to take another swing at "digest merging" from this angle: the public API in this PR should merge cleanly with that though.

Reply:
I'm also fine with abandoning this PR. It was really more of a concept PR than anything else. Thoughts?